### DISCUSSION TITLE
Using IPython

### DISCUSSION THREAD

**Discussion by sandorex** [pyodide/pyodide#2313] [Q&A] [answered]
I want to use pyodide to host my python app online with REPL, but i cannot run IPython as `termios` cannot be imported
```
...
ModuleNotFoundError: No module named 'termios'
...
```

Is there some workaround, i don't even know why termios is included at all

```py
<!DOCTYPE html>
<html>
  <head>
      <script src="https://cdn.jsdelivr.net/pyodide/v0.19.1/full/pyodide.js"></script>
  </head>
  <body>
    Pyodide test page <br>
    Open your browser console to see Pyodide output
    <script type="text/javascript">
      async function main() {
        let pyodide = await loadPyodide({
          indexURL : "https://cdn.jsdelivr.net/pyodide/v0.19.1/full/"
        });
        await pyodide.loadPackage("micropip");
        pyodide.runPythonAsync(`
            import micropip
            await micropip.install('IPython')
            import IPython

            x = IPython.embed()
            print('done', x)
        `);
      }
      main();
    </script>
  </body>
</html>
```

**Comment by hoodmane** [comment, marked as answer]
It looks like IPython needs significant patching in order to work with Pyodide. IPython normally tries to do a bunch of stuff with subprocesses etc which don't work in Pyodide.

**Reply by sandorex** [reply]
Thanks for answering, i guess i have to rewrite it and use `code.InteractiveInterpreter` or `code.InteractiveConsole`, hopefully that works

**Reply by hoodmane** [reply]
You may also be interested in [PyodideConsole](https://pyodide.org/en/stable/usage/api/python-api.html#pyodide.console.PyodideConsole) which is similar to code.InteractiveConsole but also supports top level await and loads certain packages that the user tries to import on the fly.

**Reply by sandorex** [reply]
Thanks that helps a lot!

**Comment by hoodmane** [comment]
@bollwyvl Do you have any extra info on IPython in Pyodide?

**Comment by jtpio** [comment]
For reference JupyterLite has support for IPython in the Pyolite kernel: https://github.com/jupyterlite/jupyterlite/tree/main/packages/pyolite-kernel/py/pyolite/pyolite.

**Comment by bollwyvl** [comment]
Right: we implement key parts of the jupyter kernel infrastructure, usually
provided by the ipykernel package:

https://github.com/jupyterlite/jupyterlite/blob/main/packages/pyolite-kernel/py/ipykernel/ipykernel/comm.py

We specifically mock some stdlib packages:

https://github.com/jupyterlite/jupyterlite/blob/main/packages/pyolite-kernel/py/pyolite/pyolite/mocks.py

Finally, we have some tools that wrap micropip for taking care of some of
this:

https://github.com/jupyterlite/jupyterlite/blob/main/packages/pyolite-kernel/py/piplite/piplite/piplite.py

All of this is more or less "accidental complexity" as we try to balance :
- maintainability
- enabling upstream interactive features, e.g. the ecosystem of jupyter/lab
extensions
- downstream customizability without requiring writing js, though in the
case of custom kernels it can't be avoided yet

It really depends on your use case what will work best if this even makes
sense for your
